### PR TITLE
Port mcuboot to Mbed-OS

### PR DESCRIPTION
**Pull request by AGlass0fMilk**
This PR introduces support for Mbed-OS with mcuboot. You can find demos of [a bootloader here](https://github.com/AGlass0fMilk/mbed-mcuboot-demo) and [a bootable application](https://github.com/AGlass0fMilk/mbed-mcuboot-blinky) here.

### CODE REVIEW

**Comment by douardda** [comment]
hi, nice work. I'm trying to compile the example app/bootloader. First pb: the git revision hashes in the mculib.lib and ep-oc-mcu.lib (bootloader) seem wrong (rebased git logs maybe?) using a NUCLEO_F411RE as target.

I have other errors but these might be related to my system (using debian bullseye, so arm gcc 7.3.1).

**Comment by AGlass0fMilk** [comment]
> hi, nice work. I'm trying to compile the example app/bootloader. First pb: the git revision hashes in the mculib.lib and ep-oc-mcu.lib (bootloader) seem wrong (rebased git logs maybe?) using a NUCLEO_F411RE as target.
> 
> I have other errors but these might be related to my system (using debian bullseye, so arm gcc 7.3.1).

Hi @douardda! Thanks for your response.

So those commit hashes are for specific branches. Not sure why you're having trouble with the ep-oc-mcu library but the mcuboot.lib points to my `mbed-port` branch that is the subject of the PR here.

Try this: 

Clone the `mbed-mcuboot-demo` repository and `cd` into it.
Execute `mbed deploy && mbed config root .`

These two steps should pull in the proper branches of each library. If not, please post the errors you are seeing.

You can also manually check out the proper branches. The feature of `ep-oc-mcu` I was using has since been merged so you may check out the `master` branch of that library. For mcuboot, make sure to add my fork as a remote (if it's not by default, `git remote add fork https://github.com/AGlass0fMilk/mcuboot.git`) and then check out the `mbed-port` branch.

You should then be able to compile with `mbed-cli` tools and follow the README instructions. 

@douardda How familiar are you with Mbed development? Let me know if you run into any other issues and I'll try to help sort it out.

**Comment by douardda** [comment]
> > hi, nice work. I'm trying to compile the example app/bootloader. First pb: the git revision hashes in the mculib.lib and ep-oc-mcu.lib (bootloader) seem wrong (rebased git logs maybe?) using a NUCLEO_F411RE as target.
> > I have other errors but these might be related to my system (using debian bullseye, so arm gcc 7.3.1).
> 
> Hi @douardda! Thanks for your response.
> 
> So those commit hashes are for specific branches. Not sure why you're having trouble with the ep-oc-mcu library but the mcuboot.lib points to my `mbed-port` branch that is the subject of the PR here.
> 
> Try this:
> 
> Clone the `mbed-mcuboot-demo` repository and `cd` into it.
> Execute `mbed deploy && mbed config root .`
> 

I believe it's more or less what I did, but let's try it again from scratch to be sure

> These two steps should pull in the proper branches of each library. If not, please post the errors you are seeing.
> 
> You can also manually check out the proper branches. The feature of `ep-oc-mcu` I was using has since been merged so you may check out the `master` branch of that library. For mcuboot, make sure to add my fork as a remote (if it's not by default, `git remote add fork https://github.com/AGlass0fMilk/mcuboot.git`) and then check out the `mbed-port` branch.
> 
> You should then be able to compile with `mbed-cli` tools and follow the README instructions.
> 
> @douardda How familiar are you with Mbed development? Let me know if you run into any other issues and I'll try to help sort it out.

I'm not very familiar with the mbed environment. I've used mbed on a few toy projects but using platformio as build system.

Now I'd like to make a custom board (replacement front-panel board for an old HP/Agilent test equipment) and I'd like to be able to easily upload firmwares. For now I've been using NUCLEO boards to write the code (so this part is handled by the stlink firmware) using mbed, so i though using mcuboot could be nice, and here am I :-)

**Comment by douardda** [comment]
> > Try this:
> > Clone the `mbed-mcuboot-demo` repository and `cd` into it.
> > Execute `mbed deploy && mbed config root .`
> 
> I believe it's more or less what I did, but let's try it again from scratch to be sure

```
david  (e) mbed   master  …/e/m/mbed-mcuboot-demo  mbed --version
1.10.4
david  (e) mbed  ~/p/e/mbed  git clone https://github.com/AGlass0fMilk/mbed-mcuboot-demo
Cloning into 'mbed-mcuboot-demo'...
remote: Enumerating objects: 108, done.
remote: Counting objects: 100% (108/108), done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 108 (delta 54), reused 84 (delta 32), pack-reused 0
Receiving objects: 100% (108/108), 631.78 KiB | 1.50 MiB/s, done.
Resolving deltas: 100% (54/54), done.
 david  (e) mbed  ~/p/e/mbed  cd mbed-mcuboot-demo/
 david  (e) mbed   master  …/e/m/mbed-mcuboot-demo  mbed deploy
[mbed] Working path "/home/david/perso/embed/mbed/mbed-mcuboot-demo" (program)
[mbed] Adding library "tinycbor" from "https://github.com/EmbeddedPlanet/tinycbor" at rev #6f4953e3c00a
[mbed] Adding library "mcuboot" from "https://github.com/AGlass0fMilk/mcuboot" at rev #aff9bd34ceb8
fatal: reference is not a tree: aff9bd34ceb8e4fb1f0697045e2f09764e738844
[mbed] ERROR: Unable to update "mcuboot" to rev #aff9bd34ceb8
---
```

When asking gh for this revision at https://github.com/AGlass0fMilk/mcuboot/commit/aff9bd34ceb8e4fb1f0697045e2f09764e738844 there is this little "This commit does not belong to any branch on this repository." comment, so I guess this rev belongs to a rebased branch or something.

The known revision (in the mbed-port branch) is 554a22c9bb077cff31d44bc520775d3230b066e4

**Comment by douardda** [comment]
Same for ep-oc-mcu, the refenced revision e482995c2a9346cbf5b78deb3b396f54c6ad75c1 is not in a known git branch, but according to the commit message, it's now d29aef377021a91e059ba132526ac8062ca9a5d8

**Comment by douardda** [comment]
Using correct revisions in lib fles, the "mbed deploy" runs ok

**Comment by douardda** [comment]
then the compile fails (I have gcc 7.3.1 on my debian buster laptop, which is more nitpicky/strict than gcc6) because a few names are not properly namespaced in the default_bd.cpp file:

```
[Error] default_bd.cpp@67,12: 'NULL' was not declared in this scope
[ERROR] '_queue.SimpleQueue' object has no attribute 'queue'
```

not sure about the SimpleQueue attribute.

**Comment by nvlsianpu** [comment]
@AGlass0fMilk What is timeline for mbed-os port contribution?
This PR looks much as Work-In-Progress (not clean commits history, code commented out).

**Comment by AGlass0fMilk** [comment]
@nvlsianpu I wanted to get feedback on the port through this PR.

I don't fully understand the multi-slot update feature that was introduced in the last year and I couldn't find any good API documentation on how some functions, eg: `flash_area_id_from_multi_image_slot`, are expected to behave.

This is why in my implementation these functions are pretty much stubbed out.

Can you highlight specific areas that you think need work in a review? I figured the best way to get answers to any questions that arise would be to open a PR and show that I have some understanding and have done some work/research of my own :-)

**Comment by utzig** [comment]
> I don't fully understand the multi-slot update feature that was introduced in the last year and I couldn't find any good API documentation on how some functions, eg: `flash_area_id_from_multi_image_slot`, are expected to behave.
> 
> This is why in my implementation these functions are pretty much stubbed out.

The Mynewt backend does not support multi-image so you could just do the same as can be seen here: https://github.com/JuulLabs-OSS/mcuboot/blob/master/boot/mynewt/flash_map_backend/src/flash_map_extended.c#L23

File: boot/mbed/src/flash_map_backend.cpp
Comment (nvlsianpu) [inline boot/mbed/src/flash_map_backend.cpp:40]: Is #else really needed - don't understand the case for it.

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (nvlsianpu) [inline boot/mbed/include/mcuboot_config/mcuboot_config.h:83]: we just switched to SINGLE_IMAGE_DFU -> SINGLE_APPLICATION_SLOT

**Comment by utzig** [comment]
I would suggest removing any line that is commented out from the PR, and do some rework on the commits. The PR's commit should not be a work log, but instead each commit should add some amount of functionality in a complete fashion, with no reworking of previous commits in later commits, etc, as a rule of thumb think something like "if we had to revert some of the commits for some reason how would we do it?", this should be easy to do if required. I have not looked carefully so I don' t know how it should be organized, but if you need help let me know and I will do a careful review.

**Comment by nvlsianpu** [comment]
flash_area_id_from_multi_image_slot()
Multi-image mean that MCUBoot supports multiple images at the same run-time. Each image has a slot from witch it can be run (the `primary_slot`) and a `second_slot` used for its new-image download (by DFU application). So this function returns `flash_area_id` for given image_index. slot = 0 -> get `flash_area_id` of `primary_slot`, slot = 1 ->  get `flash_area_id` of `secondary_slot`,  slot = 2 -> get scratch area ID.

**Comment by utzig** [comment]
@AGlass0fMilk Are you still working on this?

**Comment by AGlass0fMilk** [comment]
> @AGlass0fMilk Are you still working on this?

@utzig I am, along with a number of other things :sweat_smile: 

Every now and then my work comes back to this so I will be working on it in the next couple weeks most likely.

I submitted a PR to get feedback and I will apply the feedback next time I must work on mcuboot.

**Comment by evedon** [comment]
Hi @AGlass0fMilk, 

Nice work indeed! We've been using this port to run the demo app on a DISCO_L475VG_IOT01A board. We've updated some of the .lib, in particular to use Mbed OS 6.3 to run the demo successfully.
This is a good starting point to show MCUboot working with a subset of the config options.
If you are short of time, we can help with cleaning up the PR in order to get it in the upcoming MCUboot release.

**Comment by utzig** [comment]
Thanks for taking a look @evedon!

@AGlass0fMilk Our plan is to start working on a new release in a few weeks, would be really nice to have this merged by then; since it is basically compartmentalized under `boot/mbed` this should be fairly easy to get in, as long as the commits are cleaner, commented out lines are removed, etc.

**Comment by AGlass0fMilk** [comment]
> Hi @AGlass0fMilk,
> 
> Nice work indeed! We've been using this port to run the demo app on a DISCO_L475VG_IOT01A board. We've updated some of the .lib, in particular to use Mbed OS 6.3 to run the demo successfully.
> This is a good starting point to show MCUboot working with a subset of the config options.
> If you are short of time, we can help with cleaning up the PR in order to get it in the upcoming MCUboot release.
Hi @evedon, help with getting this cleaned up and ready to merge would be fantastic!

You can either submit PRs to my fork on this branch or I can give you write access to my fork. Let me know what works for your team. I'm glad my work has been useful to you!

Once your changes are merged in I can rebase to clean up the history and then we can move this PR forward.

**Comment by evedon** [comment]
@AGlass0fMilk We'll make a PR with clean history to your branch.

**Comment by AGlass0fMilk** [comment]
Took a few tries to get the "signed-of-by" email correct :sweat_smile: (side question: does it use the committer email or the GitHub user's primary email address when checking?)

:crossed_fingers: Hopefully the CI goes well and we can get this merged soon.

@utzig The Mbed OS team (@evedon @LDong-Arm) made some fantastic updates to my port, clarifying a lot of the code and writing up a nice README. 

Let me know if you require any changes to the squashed commit message -- the history should be pretty clean now :smile:

**Comment by utzig** [comment]
> Took a few tries to get the "signed-of-by" email correct  (side question: does it use the committer email or the GitHub user's primary email address when checking?)

It checks that both the committer and author emails of the commits match the sign-off lines, it's not related to the GH configuration.

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (utzig) [inline boot/mbed/include/mcuboot_config/mcuboot_config.h:45]: It is used for signature verification and ECIES using secp256r1 and AES encryption; RNG should not be required.

File: boot/mbed/include/flash_map_backend/flash_map_backend.h
Comment (utzig) [inline boot/mbed/include/flash_map_backend/flash_map_backend.h:129]: This is not used anymore but MCUBoot and can be removed.

File: boot/mbed/app_enc_keys.c
Comment (utzig) [inline boot/mbed/app_enc_keys.c:40]: Who is generating the key files for this port?

File: boot/mbed/mcuboot_main.cpp
Comment (utzig) [inline boot/mbed/mcuboot_main.cpp:47]: This block should be inside something like `#if (MCUBOOT_CRYPTO_BACKEND == MBEDTLS)`

File: boot/mbed/src/flash_map_backend.cpp
Comment (utzig) [inline boot/mbed/src/flash_map_backend.cpp:230]: Shouldn't it be `return area_id`?

File: boot/mbed/src/flash_map_backend.cpp
Comment (utzig) [inline boot/mbed/src/flash_map_backend.cpp:171]: This function is not used anymore, please remove.

File: boot/mbed/src/flash_map_backend.cpp
Comment (utzig) [inline boot/mbed/src/flash_map_backend.cpp:137]: I believe this could be `uint8_t buffer[read_size];` since `read_size` is the amount of data you put into the buffer.

File: boot/mbed/README.md
Comment (utzig) [inline boot/mbed/README.md:1]: This file has to be moved to `/docs/readme-mbedos.md` and a proper entry added to `/docs/index.md`; has to update the link to `mbed_lib.json` afterwards.

**Review by utzig — CHANGES_REQUESTED** [review]
Looks quite good, I don't really understand why you're making it hard to have a secondary slot, with those extra steps needed. The commit message needs to be updated to state what this does; what has changed is not really relevant since there is no previous mbed-os port to refer to while saying "key changes" bla bla bla; just write down what this commit does. I left a few suggestions for fixes but this is almost mergeable IMO.

File: boot/mbed/app_enc_keys.c
Comment (AGlass0fMilk) [reply boot/mbed/app_enc_keys.c:40]: Is there a way to disable encryption for this PR at least? I haven't tested it yet.

At some point I believe I had made changes to allow the application encryption keys to be defined by the user. They would then use the imgtool tools to generate .C versions of their encryption keys. Would this be an acceptable path forward?

Comment (utzig) [reply boot/mbed/app_enc_keys.c:40]: You don't need encryption if you decide so, but add some #error message if the user has enabled it.

Comment (AGlass0fMilk) [reply boot/mbed/app_enc_keys.c:40]: Oh, I didn't read the whole file. The key symbols are declared as externs and so can be provided by the user if they wish to enable encryption.

Can we just reword the comment? The encryption keys can be generated using imgtool the same way the signing keys are generated. There is no automated build step provided by Mbed-OS to do this.

Comment (utzig) [reply boot/mbed/app_enc_keys.c:40]: Sure, using `imgtool getpub ...`  should be fine!

**Comment by AGlass0fMilk** [comment]
That's too bad, we lost the fantastic `dadfeed` commit hash during the rebase :(
dadfeed-hash

**Comment by utzig** [comment]
> That's too bad, we lost the fantastic `dadfeed` commit hash during the rebase 

Well "lost" is a strong word when using git: https://github.com/JuulLabs-OSS/mcuboot/commit/dadfeed, but sure it won't show up on `master`!

File: docs/readme-mbed.md
Comment (utzig) [inline docs/readme-mbed.md:19]: The link should be changed from `../boot/mbed/mbed_lib.json` to `https://github.com/JuulLabs-OSS/mcuboot/tree/master/boot/mbed/mbed_lib.json`

File: docs/readme-mbed.md
Comment (utzig) [inline docs/readme-mbed.md:41]: The link should be changed from `https://mcuboot.com/mcuboot/design.html#image-trailer` to `design.md#image-trailer`

**Review by utzig — CHANGES_REQUESTED** [review]

**Review by utzig — COMMENTED** [review]
A few documentation issues...

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (LDong-Arm) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: [This header](https://github.com/mcu-tools/mcuboot/blob/master/ext/tinycrypt/lib/include/tinycrypt/ecc_platform_specific.h#L77) sets `#define default_RNG_defined 1` which tells source files trying to use `default_CSPRNG()` which is undefined. This is the reason we initially marked TinyCrypt as unsupported. Could we change it to 0?

File: boot/mbed/src/flash_map_backend.cpp
Comment (LDong-Arm) [inline boot/mbed/src/flash_map_backend.cpp:137]: We initially used a predefined array size, because Mbed OS enables `-Wvla` which warns usage of variable-sized arrays. But variable-sized array does bring in flexibility in our case.

@AGlass0fMilk In this case, we can remove the config from `mbed_lib.json` and the `#ifdef`s in this function?

File: .mbedignore
Comment (LDong-Arm) [inline .mbedignore:7]: @AGlass0fMilk 
TinyCrypt is ignored here (in `ext`), so it doesn't actually get built. Rather than ignoring the whole `ext`, please filter out its sub-directories instead.

After making TinyCrypt available (and change the signature algorithm to EC256, because RSA enables Mbed TLS no matter what crypto backend you specific in the config), I think you may get the issue I mentioned in https://github.com/mcu-tools/mcuboot/pull/791#discussion_r514984219

Comment (LDong-Arm) [reply .mbedignore:7]: And one more issue: If you compile with the Arm Compiler instead of GCC, `#include <version>` (e.g. in Mbed OS system code) will wrongly pick `ext/tinycrypt/VERSION` as `<version>` on case-insensitive file systems like those on Windows and macOS.

It's a known issue - the current Mbed build system recursively makes _every_ path an include path, including `ext/tinycrypt/` (but we only want `ext/tinycrypt/lib/include/`). There's no solution to this, until Mbed OS has officially announced CMake support which has fine-grained control over include paths - it's still in active development at the moment.

Comment (LDong-Arm) [reply .mbedignore:7]: So, to enable TinyCrypt for Mbed OS, we need to
* resolve https://github.com/mcu-tools/mcuboot/pull/791#discussion_r514984219
* restrict compilation to GCC only (and document this) until the new CMake-based build system is in place

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (utzig) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: > Could we change it to 0?

No, but I would suggest you try something like this on mcuboot_main.cpp:

```
#if (MCUBOOT_CRYPTO_BACKEND == TINYCRYPT)
/* XXX add this global definition for linking only */
int default_CSPRNG(uint8_t *dest, unsigned int size) { return 0; }
#endif

int main() 
{
    ...

    /* XXX add this close to Mbed TLS initialization */
    #if (MCUBOOT_CRYPTO_BACKEND == TINYCRYPT)
        uECC_set_rng(0);
    #endif

    ...
```

Comment (LDong-Arm) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: That's a good idea

File: boot/mbed/src/flash_map_backend.cpp
Comment (utzig) [reply boot/mbed/src/flash_map_backend.cpp:137]: With `MCUBOOT_READ_GRANULARITY` being a compile time var, and `read_size` a runtime var it makes sense that one would get warnings, so it's OK to rollback on my initial comment and use `MCUBOOT_READ_GRANULARITY` instead.

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (AGlass0fMilk) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: I will implement the changes above however I'm wondering if there are any impacts on security if TinyCrypt expects a TRNG and there is not one available. Is this workaround purely to satisfy missing symbols? From what @utzig said, mcuboot appears to use TinyCrypt in a way that it will not call TRNG functions.

File: boot/mbed/src/flash_map_backend.cpp
Comment (LDong-Arm) [reply boot/mbed/src/flash_map_backend.cpp:137]: I'm personally okay with both options. If the static array size (MCUBOOT_READ_GRANULARITY) is used, we can document (in `mbed-app.json`) it should be at least the value of `<blockdevice>->get_read_size()`.

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (LDong-Arm) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: > I will implement the changes above however I'm wondering if there are any impacts on security if TinyCrypt expects a TRNG and there is not one available

MCUboot only calculates checksum, where no randomness is involved, so there should be no security implication as far as I understand. RNG is more useful for key generation which is not done by our bootloader.

Comment (utzig) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: I did some re-check of the Tinycrypt code. For signature validation the TRNG really is not needed as one would expect, so no issues; for encryption it is used to blind the montgomery ladder, so it has security implications, because not using it would allow someone to use differential power analysis to recover a secret when doing the ECDH. But honestly this is just from a fast inspection, I have to think this more thoroughly; unfortunately today I don't have the time. If you are not using encrypted images you are totally fine!

Comment (LDong-Arm) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: Integration with Mbed OS TRNG should be fairly simple, the API is https://github.com/ARMmbed/mbed-os/blob/master/hal/include/hal/trng_api.h
But it can be a future addition in a separate PR, if we want to enable encryption at some point.

File: boot/mbed/src/flash_map_backend.cpp
Comment (evedon) [reply boot/mbed/src/flash_map_backend.cpp:137]: We prefer compile time var, hence why Mbed OS enables `-Wvla`. So I prefer the original option with predefined array size.

Comment (AGlass0fMilk) [reply boot/mbed/src/flash_map_backend.cpp:137]: I prefer the static array size implementation. I will update the `mbed_lib.json` help for the `MCUBOOT_READ_GRANULARITY` setting with the above information.

File: boot/mbed/include/mcuboot_config/mcuboot_config.h
Comment (AGlass0fMilk) [reply boot/mbed/include/mcuboot_config/mcuboot_config.h]: I will make sure to add a note regarding this in the PR commit message.

**Comment by AGlass0fMilk** [comment]
I have too many emails :confounded: 

Anyways, I think I got this last one right. I was able to build with the crypto backend set to both mbedtls and tinycrypt with GCC.

Let me know if there are any further comments. :crossed_fingers:

**Review by LDong-Arm — APPROVED** [review]
Tested ECDSA with TinyCrypt and RSA with Mbed TLS, both work! Thanks a lot @AGlass0fMilk @utzig 

The [demo|https://github.com/AGlass0fMilk/mbed-mcuboot-demo/pull/5] is largely in progress despite the README pointing to it, so it would be good to get that updated soon. @AGlass0fMilk

**Review by nvlsianpu — APPROVED** [review]

**Review by d3zd3z — APPROVED** [review]

**Comment by AGlass0fMilk** [comment]
All green!

@LDong-Arm 
> Tested ECDSA with TinyCrypt and RSA with Mbed TLS, both work! Thanks a lot @AGlass0fMilk @utzig
> 
> @AGlass0fMilk The [demo](https://github.com/AGlass0fMilk/mbed-mcuboot-demo/pull/5) is largely in progress despite the README pointing to it, so it would be good to get that updated soon.

Yes I will be integrating mcuboot into some projects this week so I will revise the demos then.

**Comment by evedon** [comment]
Thanks everyone

**Pull request merged**
